### PR TITLE
fix(ci): isolate GIT_* for every count-ratchet git subprocess

### LINKED ISSUE #4914: count_ratchet inherits GIT_DIR from a linked-worktree push, so the counters read the wrong tree and block the push
A `git push` that originates from a linked worktree exports `GIT_DIR` into the pre-push hook environment. A push from a normal clone does not. `scripts/ci/count_ratchet.py::tracked_files()` (around line 66) runs `git ls-files` with the ambient environment, so the exported `GIT_DIR` overrides the `-C <scratch>` argument and the counter gets the main worktree file list instead of the merged scratch tree.

Effect: any file that exists on the branch but was deleted on `origin/main` is handed to ruff (or to `open()`) as a scratch path that does not exist. The counter returns None and `merge-tree-ratchet` plus `pre-pr-validation` fail the push.

Observed on PR #4912, branch `docs/session-14684-pr-autofix-batch-d`, five consecutive blocked pushes:

```
ruff could not read <scratch>/scripts/validation/check_unreachable_after_terminator.py: No such file or directory
counter returned None
```

`scripts/validation/check_unreachable_after_terminator.py` exists on that branch and is deleted on `origin/main`.

Proof of the mechanism, run directly:

```
# from the linked worktree gitdir
GIT_DIR=<worktree gitdir> git -C <scratch> ls-files -- scripts/validation/check_unreachable_after_terminator.py
scripts/validation/check_unreachable_after_terminator.py

# clean environment, same scratch tree
git -C <scratch> ls-files -- scripts/validation/check_unreachable_after_terminator.py
(no output)
```

Also confirmed with a throwaway repo pair that git exports `GIT_DIR` to pre-push hooks only from a linked worktree.

Workaround used: push from a non-linked clone. All hooks still ran and passed there (`ruff 27 <= 27`, `type-ignore 44 <= 44`).

Fix: `scripts/ci/merge_tree_materialization.py::isolated_git_environment()` already strips `GIT_*` for materialization and scratch init. `count_ratchet.tracked_files()` and the other counter subprocesses should use the same isolation instead of the ambient environment.

Cost of not fixing: every agent that works in a worktree, which is the documented pattern, hits an unexplained blocked push whenever the branch carries a file main deleted. It cost five pushes and roughly 40 minutes on PR #4912 before the cause was found.

### PR DESCRIPTION
**Pull request by rjmurillo**
### What

Every git subprocess in `scripts/ci/count_ratchet.py` now runs under a new
`git_environment()` helper that strips `GIT_*` from a copy of `os.environ`,
instead of inheriting the ambient environment. The same one-line fix is applied
to the duplicate `baseline_at_ref` in `scripts/ci/ruff_count_ratchet.py`.

### Why

`git -C <root>` does not win against an exported `GIT_DIR`. A `git push` from a
linked worktree exports `GIT_DIR` into the pre-push hook; a push from an ordinary
checkout does not. Worktrees are the documented pattern here, so every agent is on
the affected path and nobody in a plain clone can reproduce it.

`merge_tree_ratchet_check` calls `current_count(<scratch>)`, so `tracked_files`
listed the *pushing worktree's* index against the *scratch* tree. Any file the
branch carries that `origin/main` deleted was then handed to ruff as a scratch path
that does not exist, the counter returned `None`, and `merge-tree-ratchet` plus
`pre-pr-validation` blocked the push while naming a file the branch legitimately
has. That cost five pushes and roughly 40 minutes on PR #4912.

## Specification References

| Type | Reference | Description |
|------|-----------|-------------|
| **Issue** | Fixes #4914 | count_ratchet inherits GIT_DIR from a linked-worktree push, so the counters read the wrong tree and block the push |
| **QA** | `.agents/qa/pr-4970-issue-4914-count-ratchet-git-env-qa.md` | Full evidence, bound to `b8727c8d32` |
| **Session** | `.agents/sessions/2026-08-13-session-14696-...json` | Session 14696 log |

### The mechanism, measured on git 2.43.0

Scratch repo tracking only `keep.py`; `GIT_DIR` points at a worktree that also
tracks `scripts/validation/only_on_branch.py`:

```text
$ git -C <scratch> ls-files                          # clean environment
keep.py

$ GIT_DIR=<worktree gitdir> git -C <scratch> ls-files
keep.py
scripts/validation/only_on_branch.py                 # the WRONG tree
```

Git exports the variable only from a linked worktree. Measured by dumping the
pre-push hook environment on both pushes:

| Push origin | `GIT_DIR` in the hook environment |
|-------------|-----------------------------------|
| Linked worktree | `<main>/.git/worktrees/<name>` |
| Ordinary checkout | absent |

### The issue's own failure, end to end

`ruff_count_ratchet.current_count(<scratch>)` with `GIT_DIR` pointing at a
disposable repo that carries a file the scratch tree does not. Same command,
isolation removed then restored:

```text
before: ruff could not read <scratch>/scripts/validation/
        check_unreachable_after_terminator.py: No such file or directory (os error 2)
        current_count(scratch) -> None

after:  current_count(scratch) -> 0
```

That `None` is the `counter returned None` in the issue report.

### This PR was pushed from a linked worktree

`merge-tree-ratchet` passed on the push (20.82s), which is the gate the issue
reports as blocked.

## Changes

- `scripts/ci/count_ratchet.py`: adds `git_environment()`; `tracked_files` and
  `baseline_at_ref` now delegate to the existing `_git_run`, cutting git launch
  sites from four to two, both passing `env=`. No change to behaviour or stderr
  text.
- `scripts/ci/ruff_count_ratchet.py`: same isolation on its duplicate
  `baseline_at_ref`, a `git show` against a caller-supplied root in the same
  ratchet family.
- `tests/ci/test_count_ratchet_git_environment.py`: 17 new tests.
- `.serena/memories/ci/git-dir-from-a-linked-worktree-push-outranks-git-c.md`:
  the symptom, the strip, and the trap for repro scripts. Indexed.

### Why the canonical helper was mirrored, not imported

`merge_tree_materialization.isolated_git_environment` is the existing pattern.
Two measured reasons it is not imported, both recorded in the new docstring:

1. `count_ratchet` is stdlib-only by contract (the gates run by path in CI) and
   that module imports `scripts.cli_exec`.
2. It also blanks `HOME` and the global git config. These gates run git against
   the real checkout, where `actions/checkout` records `safe.directory` globally.
   Measured: a global `safe.directory` that `git config --get-all` returns under
   the ambient environment (rc 0) is invisible under that helper's environment
   (rc 1).

So the `GIT_*` half of its rule is mirrored verbatim, `name.upper()` included,
and nothing else is dropped. The divergence is documented in a
`Stricter/looser/different than canonical` paragraph in the helper.

## Type of Change

- [x] Bug fix (non-breaking change fixing an issue)
- [ ] New feature (non-breaking change adding functionality)
- [ ] Breaking change (fix or feature causing existing functionality to change)
- [ ] Documentation update
- [x] Infrastructure/CI change
- [ ] Refactoring (no functional changes)

## Reviewer Expectations

**Review kind / scrutiny / urgency**: targeted, standard scrutiny, no rush.

**Focus areas**:

1. The scope of the strip. All `GIT_*` is stripped, matching canonical. Every
   call site is a local read-only command (`ls-files`, `diff`, `show`, `ls-tree`,
   `rev-parse`), so `GIT_SSH_COMMAND` and `GIT_ASKPASS` are not load-bearing, and
   `GIT_EXEC_PATH` falls back to git's compiled-in path.
2. The `tracked_files` and `baseline_at_ref` delegation to `_git_run`. It is
   intended to be behaviour-preserving, including stderr text. `_diff_paths` was
   deliberately left separate because its "diagnostic ordering degraded" prefix
   differs.
3. Whether fixing the `ruff_count_ratchet` duplicate belongs here. It is called
   only by its own test. It was fixed rather than deleted; deleting it is noted
   as a follow-up.

## Testing

Deliverables in this PR:

- [x] Tests added/updated
- [x] Manual testing completed
- [ ] No testing required (documentation only)

17 tests in `tests/ci/test_count_ratchet_git_environment.py`, run against real
git rather than a `subprocess.run` stand-in. A stand-in would assert that some
dict was passed and would pass just as happily if git ignored it, which is the
whole question.

| Class | Count | What it pins |
|-------|-------|--------------|
| Positive | 6 | Correct listing with no `GIT_*` set; a clean environment passes through unchanged; `PATH`, `HOME` and a `DIGIT_COUNT` lookalike survive |
| Negative | 5 | Foreign `GIT_DIR` must not reach `tracked_files`, `baseline_at_ref`, `baseline_absent_at_ref`, `changed_files`, or the `ruff_count_ratchet` duplicate |
| Edge | 4 | Lowercase `git_dir`; result is a copy not a view; a non-repository still returns `None` rather than `[]`; launch failure still returns `None` |
| Guard | 2 | AST check that no git `subprocess.run` in either module omits `env=`, plus the guard's own negative control |

Two deserve naming. `test_foreign_git_dir_would_win_without_isolation` asserts
the override is still real, so the negative controls cannot start passing for an
unrelated reason. `test_tracked_files_still_reports_failure_under_isolation`
pins that isolation does not turn a git failure into an empty listing, which
would score a tree as zero violations and pass by failing.

**Verified the tests detect the defect**: with the isolation reverted, 8 of the
17 fail, including the guard on both modules. Restored: 17 pass.

| Gate | Result |
|------|--------|
| `uv run pytest tests/ci/ -q` | 2416 passed, 10 skipped |
| `uv run python scripts/validation/pre_pr.py` | All validations passed |
| ruff / taste / type-ignore / memory-index / subprocess-encoding / cli-exit ratchets | all OK |
| Pre-push hooks (from a linked worktree) | all green, `merge-tree-ratchet` included |

### Security Review

- [x] Security agent reviewed infrastructure changes

This touches CI gate scripts, so it is flagged rather than waved through. The
change narrows what a subprocess inherits; it does not widen a boundary, add an
input path, or touch credentials. Worth a reviewer confirming point 1 under
Focus areas.

**Files requiring security review:**

- `scripts/ci/count_ratchet.py`
- `scripts/ci/ruff_count_ratchet.py`

### Other Agent Reviews

- [ ] Architect reviewed design changes
- [ ] Critic validated implementation plan
- [x] QA verified test coverage

## Author Pre-flight

- [x] Builds locally (or N/A)
- [x] Pre-PR validation passed (`uv run python scripts/validation/pre_pr.py`)
- [x] Lint and formatting clean (scoped to changed files)
- [x] Code follows project style guidelines and code quality standards
- [x] Self-review completed (read the diff as if you did not write it)
- [x] Comments added only where the *why* is non-obvious
- [x] Documentation updated (if applicable)
- [x] No new warnings introduced

## Notes for Reviewers

One thing worth flagging, because it is the best evidence that this defect is
easy to walk into. While diagnosing it, a throwaway repro script ran
`git add -A` and `git commit` with `GIT_DIR` still exported and its cwd inside a
scratch directory. Git used `GIT_DIR` as the repository and the scratch dir as
the work tree, so `add -A` staged every real file as deleted and the commit
landed on this branch, through the very defect under diagnosis. It was recovered
with `git reset --mixed`, the working tree untouched, and the branch history here
is clean: three commits, no `snap`. Both the incident and the two rules that
prevent it are recorded in the Serena memory this PR adds.

## Related Issues

Fixes #4914

### CODE REVIEW

File: .agents/sessions/2026-08-13-session-14696-b8dd33c5e-fix-issue-4914-isolate-git_.json [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      "action": "Added git_environment() to count_ratchet and routed every git subprocess through it",
+      "outcome": "tracked_files and baseline_at_ref now delegate to _git_run; git launch sites cut from 4 to 2, both passing env=. Same fix applied to the duplicate ruff_count_ratchet.baseline_at_ref"
+    },
+    {
+      "action": "Wrote tests/ci/test_count_ratchet_git_environment.py against real git",
+      "outcome": "17 tests: positive, foreign-GIT_DIR negative control, a test pinning the override itself so the control cannot pass vacuously, edge cases, and an AST guard with its own negative control"
+    },
+    {
+      "action": "Verified the tests detect the defect by reverting the isolation and re-running",
+      "outcome": "8 of 17 failed with isolation removed, including the guard on both modules; all 17 pass with it restored"
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .agents/sessions/2026-08-13-session-14696-b8dd33c5e-fix-issue-4914-isolate-git_.json]: Fixed in f3abfbdd1. Rephrased to "8 failed out of 17" so the leading integer is the failure count, not the total.

File: .agents/qa/pr-4970-issue-4914-count-ratchet-git-env-qa.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
+---
+qaVerdict: PASS
+qaSessionLog: .agents/sessions/2026-08-13-session-14696-b8dd33c5e-fix-issue-4914-isolate-git_.json
+qaCommit: b8727c8d32ae895e1dec6027d77c718271e44a6f
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .agents/qa/pr-4970-issue-4914-count-ratchet-git-env-qa.md:4]: Fixed in 816b1f8f7. The session endingCommit was updated to b8727c8d3 to match qaCommit. The mismatch was a transient state between two commits.

**Comment by rjmurillo** [comment]
owner: local:pr-autofix
session: session-14706
acquired_at: 2026-08-15T02:28:29Z
expires_at: 2026-08-15T02:43:29Z
base_sha: 816b1f8f753235f6ac10ef4931fe14a6b266d124
claim_id: f0a69ea53b2c47ec9ba50e8e22a5dc9d

**Comment by rjmurillo** [comment]
owner: local:pr-autofix
session: session-14706
acquired_at: 2026-08-15T02:29:18Z
expires_at: 2026-08-15T02:44:18Z
base_sha: 816b1f8f753235f6ac10ef4931fe14a6b266d124
claim_id: cbc3c57ef24c424385f2ef9aff30d34c

**Comment by rjmurillo** [comment]
owner: none
session: session-14706
acquired_at: 2026-08-15T02:14:56Z
expires_at: 2026-08-15T02:14:56Z
base_sha: 816b1f8f753235f6ac10ef4931fe14a6b266d124
target_owner: local:pr-autofix
claim_id: cbc3c57ef24c424385f2ef9aff30d34c
